src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs

```diff
@@ -83,6 +81,11 @@ public static partial class JsonSerializer
 
                     if (options.DictionaryKeyPolicy != null && state.Current.ExtensionDataStatus != ExtensionDataWriteStatus.Writing)
                     {
+                        if (key == null)
```

Comment:
They `key` won't be null here. This can be replaced by a `Debug.Assert(key != null);`

The check for `(key == null) throw` needs to happen after the call to `ConvertName`.

We should add such a test case. Probably something like:
```C#
public class BadNamingPolicy : JsonNamingPolicy
{
    public override string ConvertName(string name)
    {
        return null;
    }
}

[Fact]
public static void BadNamingPolicy_ThrowsInvalidOperation()
{
    var options = new JsonSerializerOptions { DictionaryKeyPolicy = new BadNamingPolicy() };

    var inputPrimitive = new Dictionary<string, int>
    {
        { "validKey", 1 }
    };

    Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(inputPrimitive, options));

    var inputClass = new Dictionary<string, Foo>
    {
        { "validKey", new Foo() }
    };

    // This assert currently fails, since we incorrectly return a JsonException
    Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(inputClass, options));
}
```

cc @Jozkee

Reply:
I see, I thought as `ConvertName(name)` will not return null unless parameter `name` is null, but yeah other overwritten method might return null.